File: app/app/src/main/java/com/kaiprompt/app/Crypto.kt
package com.kaiprompt.app

import org.json.JSONObject
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * The other half of lib/crypto.mjs on the PC.
 *
 * Everything the server sends over the Cloudflare tunnel arrives sealed, because Cloudflare
 * terminates the TLS and could otherwise read the lot: prompts, code, and everything Claude
 * said back. The key that opens it never went through the tunnel — it came off the PC's own
 * screen, in the pairing QR. That is the whole reason the tunnel is acceptable.
 *
 * AES-256-GCM: decryption fails outright on a wrong key or a tampered payload, rather than
 * quietly handing back garbage. If Cloudflare changed a byte, we find out.
 */
object Crypto {

    private const val ALGO = "AES/GCM/NoPadding"
    private const val TAG_BITS = 128

    // java.util.Base64, not android.util.Base64: it exists from API 26 (our minSdk) AND on
    // a plain JVM, which is what lets every one of these tests run without an emulator. The
    // Android one would have made this class untestable on the machine that builds it.
    private val url64: Base64.Decoder = Base64.getUrlDecoder()
    private val std64: Base64.Decoder = Base64.getDecoder()

    /** The pairing QR carries the key base64url-encoded, exactly as Node minted it. */
    fun keyBytes(key: String): ByteArray = url64.decode(key)

    /**
     * A sealed envelope → the JSON inside.
     * Throws if the key is wrong or the payload was touched — never returns something we
     * cannot vouch for.
     */
    fun open(envelope: String, key: String): String {
        val json = JSONObject(envelope)
        require(json.optInt("v") == 1) { "not a sealed payload" }

        val iv = std64.decode(json.getString("iv"))
        val ct = std64.decode(json.getString("ct"))
        val tag = std64.decode(json.getString("tag"))

        // Java's GCM expects the tag appended to the ciphertext; Node keeps it separate.
        // Joining them here is what makes the two sides speak the same language.
        val cipher = Cipher.getInstance(ALGO)
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(keyBytes(key), "AES"),
            GCMParameterSpec(TAG_BITS, iv),
        )
        return String(cipher.doFinal(ct + tag), Charsets.UTF_8)
    }

    /** Is this a sealed envelope, or plain JSON? The server answers plainly to anyone who did not ask. */
    fun isSealed(body: String): Boolean = try {
        val j = JSONObject(body)
        j.optInt("v") == 1 && j.has("iv") && j.has("ct") && j.has("tag")
    } catch (_: Exception) {
        false
    }
}

File: lib/crypto.mjs
// End-to-end encryption between this machine and the phone.
//
// The tunnel goes through Cloudflare, which terminates TLS — so without this, Cloudflare
// could read your prompts, your code and everything Claude said back. With it, they move
// bytes they cannot read.
//
// The key never crosses the tunnel. It is generated here and handed to the phone in the
// pairing QR, which you scan off your own screen, so it goes straight from this machine to
// that phone and nowhere else. That is what makes "nothing readable leaves your machine"
// true again even though the wire belongs to someone else.
//
// AES-256-GCM: encrypts AND authenticates, so a tampered payload is rejected rather than
// silently decrypted into garbage.

import crypto from 'node:crypto';

const ALGO = 'aes-256-gcm';
const IV_BYTES = 12;                      // GCM's nonce size

export const newKey = () => crypto.randomBytes(32).toString('base64url');
const keyBuf = (key) => Buffer.from(String(key), 'base64url');

/** Anything JSON-serialisable → a sealed envelope. */
export function seal(value, key) {
  const iv = crypto.randomBytes(IV_BYTES);
  const cipher = crypto.createCipheriv(ALGO, keyBuf(key), iv);
  const ct = Buffer.concat([cipher.update(JSON.stringify(value), 'utf8'), cipher.final()]);

  return {
    v: 1,
    iv: iv.toString('base64'),
    ct: ct.toString('base64'),
    tag: cipher.getAuthTag().toString('base64'),   // proves nobody edited it in transit
  };
}

/**
 * Envelope → the value. Throws if the key is wrong OR the payload was touched: GCM will
 * not hand back plaintext it cannot vouch for, and neither will we.
 */
export function open(envelope, key) {
  if (!envelope || envelope.v !== 1) throw new Error('not a sealed payload');

  const decipher = crypto.createDecipheriv(ALGO, keyBuf(key), Buffer.from(envelope.iv, 'base64'));
  decipher.setAuthTag(Buffer.from(envelope.tag, 'base64'));

  const plain = Buffer.concat([
    decipher.update(Buffer.from(envelope.ct, 'base64')),
    decipher.final(),                      // throws on a bad key or a tampered payload
  ]);
  return JSON.parse(plain.toString('utf8'));
}

/** Did the client ask for a sealed answer? The app always does; curl and tests may not. */
export const wantsSealed = (req) =>
  req.headers['x-kaip-enc'] === '1' || new URL(req.url, 'http://x').searchParams.get('enc') === '1';
